### PR TITLE
Optionally bring your own WorkerPool

### PR DESCRIPTION
**Pull request by imikushin**
Optionally bring your own WorkerPool (and event processing locker - as part of it)

Also:
- stop pongHandler when stopping the event router
- don't panic on unexpected connection loss

### CODE REVIEW

**Comment by cjellick** [comment]
@imikushin can you rebase to kickoff a new build with the new drone config?

**Comment by imikushin** [comment]
sure!

**Comment by cjellick** [comment]
golint errors @imikushin

**Comment by imikushin** [comment]
... and it's fixed!

File: events/listener.go [hunk 1]
```diff
@@ -33,8 +35,15 @@ type EventRouter struct {
 	eventStream   *websocket.Conn
 	resourceName  string
 	pingConfig    PingConfig
+	eventLocker   EventLocker
 }
 
+type EventLocker func(event *Event) locks.Locker
+
+func nopLocker(_ *Event) locks.Locker { return locks.NopLocker() }
```
Comment (cjellick) [inline events/listener.go]: just curious: why the underscore for the receive variable?

File: events/listener.go [hunk 2]
```diff
@@ -169,16 +184,21 @@ func (router *EventRouter) subscribeToEvents(subscribeURL string, accessKey stri
...
-			"statusCode":      resp.StatusCode,
-			"responseHeaders": resp.Header,
-			"responseBody":    string(body[:]),
-			"error":           err,
-			"subscribeUrl":    subscribeURL,
-		}).Error("Failed to subscribe to events.")
+			"subscribeUrl": subscribeURL,
+		}).Errorf("Error subscribing to events: %s", err)
+		if resp != nil {
+			log.WithFields(log.Fields{
```
Comment (cjellick) [inline events/listener.go:179]: FYI, I've come to dislike log.WithFields and prefer just normal `*Level*F` log methods. So, in the future, feel free to use those, don't need to keep with the bad precedent i set.

No need to change it in this this PR now though.

File: events/worker.go [hunk 3]
```diff
@@ -1,40 +1,44 @@
...
+			defer func() { wp <- w }()
+			doWork(message, eventHandlers, apiClient, locker)
+		}()
+	default:
 		log.WithFields(log.Fields{
-			"err": err,
-		}).Error("Error unmarshalling event")
-		return
+			"workerCount": cap(wp),
+		}).Info("No workers available dropping event.")
```
Comment (cjellick) [inline events/worker.go]: I think this is important and abnormal enough to be logged at warn level. Can you make it warn?

File: locks/locks.go [hunk 4]
```diff
@@ -1,22 +1,49 @@
...
+type Locker interface {
+	Lock() Unlocker
+}
+
 // Unlocker Interface for unlocking a key that was provided to Lock()
 type Unlocker interface {
 	Unlock()
 }
 
+type KeyLockerImpl struct {
```
Comment (cjellick) [inline locks/locks.go]: I find it confusing that KeyLockerImpl is declared here but then nop is declared and all its methods follow and then we jump back to keylocker. can you rearrange such that keylockerimpl is near its functions?

File: locks/locks.go [hunk 5]
```diff
@@ -1,22 +1,49 @@
...
+}
+
+func KeyLocker(key interface{}) *KeyLockerImpl {
+	return &KeyLockerImpl{key: key}
+}
+
 // Lock Provides an application-wide lock on the provided key.
 // If the lock is obtained, it will return an Unlocker.
 // If the lock is not successfully obtained, nil will be returned.
-func Lock(key string) Unlocker {
```
Comment (cjellick) [inline locks/locks.go:18]: The Lock api is used by other projects on its own. From either a design or impl perspective, do you see any harm in keeping a backwards compatibel version of this API in addition to what you've added?

File: events/listener.go (hunk 1 above)
Comment (imikushin) [inline events/listener.go]: it's not used in the method, so doesn't need to be named. I can even drop the underscore (just found out)

File: events/worker.go (hunk 3 above)
Comment (imikushin) [inline events/worker.go]: Sure

File: locks/locks.go (hunk 5 above)
Comment (imikushin) [inline locks/locks.go:18]: Nope, no harm at all. Adding this back:

``` go
func Lock(key interface{}) Unlocker {
    return KeyLocker(key).Lock()
}
```

**Comment by imikushin** [comment]
I believe I've addressed the comments

File: events/listener.go [hunk 6]
```diff
@@ -90,12 +100,16 @@ func (router *EventRouter) StartWithoutCreate(ready chan<- bool) error {
 	return router.run(ready, "")
 }
 
-func (router *EventRouter) run(ready chan<- bool, eventSuffix string) (err error) {
-	workers := make(chan *Worker, router.workerCount)
-	for i := 0; i < router.workerCount; i++ {
-		w := newWorker()
-		workers <- w
+func (router *EventRouter) StartNoEH(eventLocker EventLocker) error {
```
Comment (cjellick) [inline events/listener.go]: What does EH stand for?

**Comment by imikushin** [comment]
Ping @cjellick

**Comment by cjellick** [comment]
LGTM

**Pull request merged**